### PR TITLE
demo: lvgl: Add functionality to draw touch events on the screen

### PR DESCRIPTION
**Pull request by MrMarteng**
Enabling kind of a drawing feature to the LVGL hello world app.
This can be useful to test touch input functionality during board
setup.

### CODE REVIEW

**Comment by MrMarteng** [comment]
LVGL memory pool size of lvgl demo has been increased from 16k to 32k as the canvas layer requires extra memory. I couldn't manage to scale the mempool size depending on the drawing feature to be enabled.

**Comment by kartben** [comment]
> LVGL memory pool size of lvgl demo has been increased from 16k to 32k as the canvas layer requires extra memory. I couldn't manage to scale the mempool size depending on the drawing feature to be enabled.

Ya I think this is maybe too much to have this always on? Maybe the "drawing" feature can be added via an overlay? 
Also worth noting the recently added HEAP_MEM_POOL_ADD_SIZE_ construct that could probably be used here? https://docs.zephyrproject.org/latest/kernel/memory_management/heap.html#defining-the-heap-memory-pool

**Comment by kartben** [comment]
But in any case it feels like we need to go back to having an *actual* **basic** LVGL sample, with minimal footprint and virtually just a hello world label (maybe not even a button? i.e just enough to demonstrate the minimal code/config needed to get LVGL to run).

**Comment by MrMarteng** [comment]
> > LVGL memory pool size of lvgl demo has been increased from 16k to 32k as the canvas layer requires extra memory. I couldn't manage to scale the mempool size depending on the drawing feature to be enabled.
> 
> Ya I think this is maybe too much to have this always on? Maybe the "drawing" feature can be added via an overlay? Also worth noting the recently added HEAP_MEM_POOL_ADD_SIZE_ construct that could probably be used here? https://docs.zephyrproject.org/latest/kernel/memory_management/heap.html#defining-the-heap-memory-pool

For that functionality a canvas layer is needed, having the full size of the screen. I already minimized the size to 2bpp. But still it requires additional memory.

**Comment by MrMarteng** [comment]
> But in any case it feels like we need to go back to having an _actual_ **basic** LVGL sample, with minimal footprint and virtually just a hello world label (maybe not even a button? i.e just enough to demonstrate the minimal code/config needed to get LVGL to run).

I agree. So maybe moving this functionality to a separate sample to test LVGL touch functionality?

File: samples/subsys/display/lvgl/README.rst [hunk 1]
```diff
@@ -61,6 +68,10 @@ or
 - :ref:`mimxrt1060_evk`
 - `RK043FN02H-CT`_
 
+Memory requirements
+===================
+The sample applications requires a LVGL memory pool of at least 32kBytes.
```
Comment (kartben) [inline samples/subsys/display/lvgl/README.rst:73]:
```suggestion
The sample application requires an LVGL memory pool of at least 32 kBytes.
```

**Review by kartben — CHANGES_REQUESTED** [review]
some early doc feedback

File: samples/subsys/display/lvgl/README.rst [hunk 2]
```diff
@@ -44,7 +51,7 @@ for Arduino connectors, for example:
 
 or a board with an integrated display:
 
-- :ref:`esp_wrover_kit`
+- :ref:`esp_wrover_kit` and :ref:`m5stack_core2`
```
[the remark this answers is not in the document]
Comment (MrMarteng) [reply samples/subsys/display/lvgl/README.rst]: esp_wrover_kit, at least in the version referenced here, actually features a display (https://docs.zephyrproject.org/latest/boards/xtensa/esp_wrover_kit/doc/index.html)

But you are absolutely right. I corrected the description in https://github.com/zephyrproject-rtos/zephyr/commit/b371395dae3ffc7611c938516bb2840399fd54d4

File: samples/subsys/display/lvgl/Kconfig [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Martn Kiepfer <mrmarteng@teleschirm.org>
+# SPDX-License-Identifier: Apache-2.0
+
+menu "LVGL basic sample"
+
+	config SAMPLE_LVGL_TOUCH_DRAWING
+		bool "Enable touch drawing feature"
+		depends on LV_Z_POINTER_INPUT
+		help
+		This allows test of touch functionaliy by drawing all recogognized touch events on the screen.
```
[the remark this answers is not in the document]
Comment (MrMarteng) [reply samples/subsys/display/lvgl/Kconfig]: fixed in https://github.com/zephyrproject-rtos/zephyr/commit/b371395dae3ffc7611c938516bb2840399fd54d4

File: samples/subsys/display/lvgl/README.rst [hunk 4]
```diff
@@ -16,6 +16,13 @@ additional widgets may be displayed and additional interactions enabled:
       If your board has a touch panel controller
       (:dtcompatible:`zephyr,lvgl-pointer-input`), a button widget is displayed
       in the center of the screen. Otherwise a label widget is displayed.
+      The demo also features drawing functionality, which needs to be enabled
+      via menuconfig (`SAMPLE_LVGL_TOUCH_DRAWING`). When enabled, received touch
+      input will be drawn onto the surface of the screen. By clicking the button
+      widget at the centor of the screen, the screen will be cleared again. This
```
[the remark this answers is not in the document]
Comment (MrMarteng) [reply samples/subsys/display/lvgl/README.rst]: fixed in https://github.com/zephyrproject-rtos/zephyr/commit/b371395dae3ffc7611c938516bb2840399fd54d4

File: samples/subsys/display/lvgl/README.rst [hunk 5]
```diff
@@ -16,6 +16,13 @@ additional widgets may be displayed and additional interactions enabled:
       If your board has a touch panel controller
       (:dtcompatible:`zephyr,lvgl-pointer-input`), a button widget is displayed
       in the center of the screen. Otherwise a label widget is displayed.
+      The demo also features drawing functionality, which needs to be enabled
+      via menuconfig (`SAMPLE_LVGL_TOUCH_DRAWING`). When enabled, received touch
```
[the remark this answers is not in the document]
Comment (MrMarteng) [reply samples/subsys/display/lvgl/README.rst]: fixed in https://github.com/zephyrproject-rtos/zephyr/commit/b371395dae3ffc7611c938516bb2840399fd54d4

File: samples/subsys/display/lvgl/README.rst (hunk 1 above)
Comment (MrMarteng) [reply samples/subsys/display/lvgl/README.rst]: Fixed in https://github.com/zephyrproject-rtos/zephyr/commit/800ff3dbbc54d3249f04a3522125877da76cbb6f#diff-a5534d3472f36eaab1e54778a5e0a130d17d06c6a4b83ca008da78a328e0f877

File: modules/lvgl/input/lvgl_pointer_input.c [hunk 6]
```diff
@@ -46,15 +54,6 @@ static void lvgl_pointer_process_event(const struct device *dev, struct input_ev
 		return;
 	}
 
-	/* adjust coordinates */
```
Comment (faxe1008) [inline modules/lvgl/input/lvgl_pointer_input.c:49]: Is there any benefit I fail to see for moving this up into the switch? I'd prefer the compactness of the previous state.

File: samples/subsys/display/lvgl/Kconfig [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Martn Kiepfer <mrmarteng@teleschirm.org>
+# SPDX-License-Identifier: Apache-2.0
+
+menu "LVGL basic sample"
+
+	config SAMPLE_LVGL_TOUCH_DRAWING
+		bool "Enable touch drawing feature"
+		depends on LV_Z_POINTER_INPUT
+		help
+		This allows test of touch functionality by drawing all recognized touch events on the screen.
```
Comment (faxe1008) [inline samples/subsys/display/lvgl/Kconfig:10]: Nit: indent.

File: samples/subsys/display/lvgl/src/main.c [hunk 8]
```diff
@@ -146,7 +196,7 @@ int main(void)
 
 	while (1) {
 		if ((count % 100) == 0U) {
-			sprintf(count_str, "%d", count/100U);
+			sprintf(count_str, "%d", count / 100U);
```
Comment (faxe1008) [inline samples/subsys/display/lvgl/src/main.c:199]: Nit: Please do not reformat unrelated code in the same commit.

File: samples/subsys/display/lvgl/README.rst [hunk 9]
```diff
@@ -16,6 +16,14 @@ additional widgets may be displayed and additional interactions enabled:
       If your board has a touch panel controller
       (:dtcompatible:`zephyr,lvgl-pointer-input`), a button widget is displayed
       in the center of the screen. Otherwise a label widget is displayed.
+
+      The demo also features drawing functionality, which can to be enabled
+      via menuconfig (``SAMPLE_LVGL_TOUCH_DRAWING``). When enabled, received touch
```
Comment (faxe1008) [inline samples/subsys/display/lvgl/README.rst:21]: @kartben I don't know it by heart but I think there is a way to link to the kconfig symbol `CONFIG_` instead?

File: samples/subsys/display/lvgl/src/main.c [hunk 10]
```diff
@@ -125,12 +149,38 @@ int main(void)
...
+#ifdef CONFIG_SAMPLE_LVGL_TOUCH_DRAWING
+		/**
+		 * The required canvas buf size is caluclated by
+		 * (lv_img_color_format_get_px_size(cf) * width) / 8 * height)
+		 * 2 bpp are used here to reduce overall memory requiremenets
+		 **/
+		lv_opa_t *mask_map = lv_mem_alloc(LV_VER_RES * LV_HOR_RES * sizeof(lv_opa_t) / 4U);
+
+		if (mask_map == NULL) {
+			LOG_ERR("Allocating canvas layer failed. Pleas review LVGL mem pool size.");
```
Comment (faxe1008) [inline samples/subsys/display/lvgl/src/main.c:163]: Nit: "Please"

File: samples/subsys/display/lvgl/src/main.c [hunk 11]
```diff
@@ -125,12 +149,38 @@ int main(void)
 
 	if (IS_ENABLED(CONFIG_LV_Z_POINTER_KSCAN) || IS_ENABLED(CONFIG_LV_Z_POINTER_INPUT)) {
 		lv_obj_t *hello_world_button;
+		lv_obj_t *canvas = NULL;
+
+#ifdef CONFIG_SAMPLE_LVGL_TOUCH_DRAWING
+		/**
```
Comment (faxe1008) [inline samples/subsys/display/lvgl/src/main.c:155]: Would it make the `main` method maybe a bit more readable to extract this into a separate function?

**Review by faxe1008 — CHANGES_REQUESTED** [review]
Small nits. However I am not sure if we are extending the sample too far at this point. Maybe this could be its own sample?

File: samples/subsys/display/lvgl/README.rst (hunk 9 above)
Comment (kartben) [reply samples/subsys/display/lvgl/README.rst:21]: There is (``:kconfig:option:`...` ``, but it doesn't make sense in the context of a sample-specifc Kconfig since the option does not appear in the documentation hence does not make sense linking to.
Good question though, and I have to admit I too did a double-take initially and almost suggested to add a link myself :)

**Comment by pdgendt** [comment]
> Small nits. However I am not sure if we are extending the sample too far at this point. Maybe this could be its own sample?

I agree, maybe create a "drawing sample" where the touch is used to draw an a black canvas for example.

**Comment by MrMarteng** [comment]
@kartben @faxe1008 @pdgendt 
Thanks for your feedback!

I will create a separate "lvgl touch test" sample application.

This PR is being closed. Separate PR follows ASAP.
